WIP: delimited file parsing with type conversion

tfenne opened the pull request:
DO NOT MERGE

tfenne commented:
@nh13 I think this is ready for a first pass review.  I'd welcome ideas for other things that should be tested.

A couple of deficiencies I don't want to try and fix just yet:
1. I found through trial and error that Dagr's reflective string to type doesn't support `Char` which would be nice
2. It will currently blow up if you use `row[T].apply("col")` on a row that has no value in the column and the target type can't be constructed from an empty string.  Maybe that's ok?  The workaround is to use `row[T].get("col")` which will return a None if the field is empty.

nh13 commented on src/main/scala/com/fulcrumgenomics/util/DelimitedDataParser.scala:
```diff
@@ -0,0 +1,128 @@
+  * Represents a row of parsed data.  Provides methods for accessing values in a type-safe
+  * way either via apply() methods for non-optional fields or via get for optional fields.
+  */
+class Row (private val headerIndices: Map[String,Int], private val fields: Array[String], val trim: Boolean) {
```
Should the constructor be private, since it is only used in one place?  If not, probably a good idea to provide documentation for the parameters.

tfenne replied:
I'll make it `private[util]`.  I wanted to make it private but then the parser couldn't call it.  I wanted to make `Row` an inner of the parser, but then the parser couldn't extend `Iterator[Row`.

nh13 commented on src/main/scala/com/fulcrumgenomics/util/DelimitedDataParser.scala:
```diff
@@ -0,0 +1,128 @@
+    if (trim) string.trim else string
+  }
+
+  /** Fetches a value of the desired type by 0-based column index. */
```
by the

nh13 commented on src/main/scala/com/fulcrumgenomics/util/DelimitedDataParser.scala:
```diff
@@ -0,0 +1,128 @@
+  }
+
+  /** Fetches a value of the desired type by column name. */
+  def apply[A](column: String)(implicit tag: ru.TypeTag[A]): A = apply(headerIndices.apply(column))
```
why not `apply(headerIndicies(column))`? why the extra `apply`?

tfenne replied:
Fixed.  I'd made a bunch of "exploratory" changes because the compiler was giving me an inscrutable error message at some point so I was being more explicit everywhere.

nh13 commented on src/main/scala/com/fulcrumgenomics/util/DelimitedDataParser.scala:
```diff
@@ -0,0 +1,128 @@
+    }
+    else {
+      val c = ReflectionUtil.typeTagToClass(tag)
+      Some(ReflectionUtil.constructFromString(c, c, value(index)).get.asInstanceOf[A])
```
1. `value(index)` could be `string`, or you can just call `this.apply(index)`.  
2. This needs some test coverage.

nh13 commented on src/main/scala/com/fulcrumgenomics/util/DelimitedDataParser.scala:
```diff
@@ -0,0 +1,128 @@
+  }
+
+  /** Gets a value from the specified column. If the value is null or empty returns None. */
+  def get[A](column: String)(implicit tag: ru.TypeTag[A]): Option[A] = get(headerIndices.apply(column))
```
ditto about `apply`

nh13 commented on src/main/scala/com/fulcrumgenomics/util/DelimitedDataParser.scala:
```diff
@@ -0,0 +1,128 @@
+object DelimitedDataParser {
+  val DefaultDelimiter: Char = '\t'
+  val DefaultBlankPolicy: Boolean = true
+  val DefaultTrim : Boolean = true
```
I've gotten to like aligning the equal signs, how about here?

nh13 commented on src/main/scala/com/fulcrumgenomics/util/DelimitedDataParser.scala:
```diff
@@ -0,0 +1,128 @@
+  val DefaultTrim : Boolean = true
+
+  /** Constructs a DelimitedDataParser for a path. */
+  def apply(path: FilePath, delimiter: Char): DelimitedDataParser =
```
`delimiter: Char = DefaultDelimiter`?

tfenne replied:
I left it un-defaulted because you can't have default values for two methods of the same name, so I can't put defaults on both apply methods, and I would like them to be symmetric.  Maybe I'll just remove the default delimiter everywhere.

nh13 commented on src/main/scala/com/fulcrumgenomics/util/DelimitedDataParser.scala:
```diff
@@ -0,0 +1,128 @@
+    new DelimitedDataParser(Io.toSource(path).getLines(), delimiter=delimiter)
+
+  /** Constructs a DelimitedDataParser for a path. */
+  def apply(lines: Traversable[String], delimiter: Char): DelimitedDataParser =
```
ditto about delimiter.

tfenne replied:
Cannot, sadly.

nh13 commented on src/main/scala/com/fulcrumgenomics/util/DelimitedDataParser.scala:
```diff
@@ -0,0 +1,128 @@
+  // A map of header name to column index
+  private val headerIndices: Map[String,Int] = _headers.zipWithIndex.toMap
+
+  /** Returns an iterator over the headers encountered in the file, in order. */
```
returns a `Seq`, not an `Iterator`.

tfenne replied:
Good catch, thanks.

nh13 commented on src/main/scala/com/fulcrumgenomics/util/DelimitedDataParser.scala:
```diff
@@ -0,0 +1,128 @@
+      throw new IllegalStateException(s"Incorrect number of values in line.")
+    }
+
+    new Row(headerIndices, fields, trimFields)
```
name boolean parameters please: `trim=trimFields`.

nh13 commented on src/main/scala/com/fulcrumgenomics/util/DelimitedDataParser.scala:
```diff
@@ -0,0 +1,128 @@
+                          val delimiter: Char = DelimitedDataParser.DefaultDelimiter,
+                          val ignoreBlankLines: Boolean = DelimitedDataParser.DefaultBlankPolicy,
+                          val trimFields: Boolean = DelimitedDataParser.DefaultTrim) extends Iterator[Row] with LazyLogging {
+
```
It may be useful if the delimiter were a regex, for example, if we we want to ignore commas within quotes we could use `"\t(?=([^\"]*\"[^\"]*\")*[^\"]*$)".r`.

nh13 commented:
@tfenne minor stuff, but I do think it would be useful to support a `Pattern` for parsing the file.

tfenne commented:
@nh13 I made most of your suggested changes.  I also implemented my own string splitting as when I added a test with just delimiters on a line, or trailing multiple delimiters together _none_ of the available methods worked (not's scala's `String.split(Char)`, not java's `String.split(Pattern)`, nor htsjdk's `StringUtil.split`).  They all returned less than the expected number of fields.

I don't think pattern based splitting is the way to go here.  I think if we want to handle quoting (which I'm deferring altogether since I don't need it currently) the right thing to do is implement some state-fulness in the splitting.  I'm not convinced that a regex can handle the full set of necessary things including quoting, escaping of quotes and escaping of escape characters properly.

nh13 commented:
👍 squash and merge at will.
